sdk/identity/azure-identity/src/token_cache.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "private/token_cache.hpp"
+#include "private/token_cache_internals.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <mutex>
+#include <utility>
+
+using Azure::Identity::_detail::TokenCache;
+
+using Azure::Core::Credentials::AccessToken;
+
+decltype(TokenCache::Internals::Cache) TokenCache::Internals::Cache;
+decltype(TokenCache::Internals::Expirations) TokenCache::Internals::Expirations;
+decltype(TokenCache::Internals::Mutex) TokenCache::Internals::Mutex;
+
+#if defined(TESTING_BUILD)
+decltype(TokenCache::Internals::OnBeforeWriteLock) TokenCache::Internals::OnBeforeWriteLock;
+#endif
+
+AccessToken TokenCache::GetToken(
+    std::string const& tenantId,
+    std::string const& clientId,
+    std::string const& authorityHost,
+    std::string const& scopes,
+    std::function<AccessToken()> const& getNewToken)
+{
+  // If cached token expires in less than MinExpiry from now, it's cached value won't be returned,
+  // newer value will be requested.
+  constexpr auto MinExpiry = std::chrono::minutes(3);
+
+  Internals::CacheKey const key{tenantId, clientId, authorityHost, scopes};
+
+  {
+    std::shared_lock<std::shared_timed_mutex> readLock(Internals::Mutex);
+
+    auto found = Internals::Cache.find(key);
+    if (found != Internals::Cache.end()
+        && found->second.ExpiresOn > std::chrono::system_clock::now() + MinExpiry)
+    {
+      return found->second;
+    }
+  }
+
+#if defined(TESTING_BUILD)
+  if (TokenCache::Internals::OnBeforeWriteLock != nullptr)
+  {
+    TokenCache::Internals::OnBeforeWriteLock();
+  }
+#endif
+
+  std::unique_lock<std::shared_timed_mutex> writeLock(Internals::Mutex);
```

Comment:
We must not hold the write lock while refreshing the token, which usually takes hundreds of milliseconds.

This would make all the requests stall a period of time.

Reply:
I got an idea.

We add a boolean flag `TokenBeingRefreshed`. The first thread observed a token expires in 3 mins, it atomically set this value to `true`. Other threads will use the old token if it sees this value is `true`, because this old token is not expired yet. This can help us avoid too many threads refreshing the same token at the same time.

If other threads see this token is about to expire in 1 minute, it refreshes the token regardless the boolean flag.

Reply:
It was not possible for multiple threads to start refreshing the same token at the same time before, but no one during that update could read their token from cache even if their token was still good, you're right.

So I changed the code in the spirit of problems that you highlighted and partly in the spirit of your suggestions. Check it out. I'll summarize in a separate message.